core/src/main/java/org/apache/iceberg/avro/PruneColumns.java

```diff
@@ -247,8 +255,9 @@ private static Schema copyRecord(Schema record, List<Schema.Field> newFields) {
   }
 
   private static Schema.Field copyField(Schema.Field field, Schema newSchema, Integer fieldId) {
+    // do not copy over non-null default values as the file is expected to have values for fields in the file schema
     Schema.Field copy = new Schema.Field(field.name(),
-        newSchema, field.doc(), field.defaultVal(), field.order());
+        newSchema, field.doc(), hasNonNullDefault(field) ? null : field.defaultVal(), field.order());
```

Comment:
Why detect non-null here? Couldn't this always pass null because it is either null already or will be replaced with null?

Reply:
I think this should be similar to field construction in the converter. Instead of using `hasNonNullDefault`, this should be based on whether the value needs a default. In the converter, we use this: `structField.isOptional() ? JsonProperties.NULL_VALUE : null`. For Avro, that would be `isOptionSchema(newSchema) ? JsonProperties.NULL_VALUE : null`.

That way any optional field gets a null default and the default is left unspecified for required fields.

Reply:
The issue is that PruneColumns#copyField can also be called at `https://github.com/apache/iceberg/blob/14fb95519b3442092eb6aa02a2608e97e2e8dfd8/core/src/main/java/org/apache/iceberg/avro/PruneColumns.java#L88`. In that case, it is possible that `newSchema` is an option schema but with the NULL type as the second option and hence `JsonProperties.NULL_VALUE` is not an appropriate default.

Reply:
If the schema is an option schema, then its default should be null. That means that we would need to reorder the options to allow that default, I think.

Reply:
:+1:  So I think this PR now simplifies to "if we have an option schema where the first option is not NULL, we should reorder it to ensure it is NULL while building the projection schema". Updated.